Migrate down fixes

gizzmo opened the pull request:
This fixes couple bugs when migrating down.

First the `find_migrations` when the array was reversed, the keys were not preserved.
Second when migrating down, the version is always one below the last migration run.

gizzmo committed:
preserve keys when reversing the migrations array.

Signed-off-by: Gizzmo <justgiz@gmail.com>

gizzmo committed:
When migrating down, the current version is always one below the last migration.

Signed-off-by: Gizzmo <justgiz@gmail.com>

philsturgeon commented:
It is not always just -1, numbers can be skipped to allow for time-based migrations.

billmn commented:
I noticed that when I run a migration with OIL, "migration_db_table"  and "config/migrations.php" are updated

But if I run a migration (up) using 

<pre>Migrate::latest()</pre>

it update only "migration_db_table" and not the config file, so it's impossible to run a migrate:down from OIL ( it says "You are already on the first migration ..." )

philsturgeon commented:
Doing latest() from within files is a bit stupid, as if you are able to edit files easily then just edit the current migration to be the newest one! :) Oil is a command line tool and as such makes your life easier by modifying your config files for you, but i dont think libraries should be overriding config values. Do you?

gizzmo commented:
well @philsturgeon i noticed that, for example, when migrating from 2 to 0 it would run migrations down of 2, then down of 1 and the database would be at version 1 and not 0 as expected. Maybe instead of  decreasing `$ver` we could just use the passed `$version`.

For the libraries overriding config values. it does make a little sense. if you don't use the migrate task you then have to rewrite functionality that's in the migrate task, and as @billmn pointed out, your config and database can get out of sync.

gizzmo commented:
I also thought about offsetting the key value of the `$migrations` array when migrating down : https://github.com/fuel/core/blob/1.1/develop/classes/migrate.php#L241

gizzmo committed:
Offset the key value of the `$migration` array

Migrating from version 3 to 2 you need to call the `down` method of migration 3, then your at verision 2.

Signed-off-by: Gizzmo <justgiz@gmail.com>

crynobone commented:
I got an issue today

```
$ php oil r migrate:up
Migrated to version: 28 for app:default.
```

and then 

```
$ php oil r migrate:down
Migrated to version: 27 for app:default.
```

FuelPHP update the version number correctly in config/migrations.php, however in the db the version goes down to 0. Opening logs show following.

```
Debug - 2011-11-10 12:26:38 --> Migrate class initialized
Info - 2011-11-10 12:26:38 --> Migrating to: 28
Info - 2011-11-10 12:26:38 --> Migrated to 28 successfully.
Debug - 2011-11-10 12:26:46 --> Migrate class initialized
Info - 2011-11-10 12:26:46 --> Migrating to: 0
Info - 2011-11-10 12:26:46 --> Migrated to 0 successfully.
```

philsturgeon commented on classes/migrate.php:
```diff
@@ -238,6 +230,7 @@ class Migrate
 			{
 				if ($end_version === null or $version <= $end_version)
 				{
+					$direction === 'down' and --$version;
```
We cannot guess which version based on incrementing or decrementing a number, it needs to be "the next in the array". Is that how this is working? If not, that's how it SHOULD be working :)

We'll be switching to a more Rails-style approach of having a timestamp so your migrations will start to have names like 201111092342_Foo and 201111082345_Bar, so no +1 or -1 will help there.

gizzmo replied:
Took me a while to figure this out, but i think i understand. In my tests iv had every migration and wasn't missing any steps.

philsturgeon replied:
Right, set up migrations as 001, 002, 003, 005, 006, 010. 

If you can navigate up and down through them all then we are ok, otherwise this is not complete. I'm not saying it works now, but I can't merge if it does not work with those :)

gizzmo committed:
Instead of decreasing the key value when migrating down, offset them by one version.

gizzmo commented:
I would like to point out that this works, but when trying to use migrate:down or migrate:up breaks when migrating to/from a version that doesnt exists. in tasks/migrate.php on [line 224](https://github.com/fuel/core/blob/1.1/develop/tasks/migrate.php#L224) and [line 249](https://github.com/fuel/core/blob/1.1/develop/tasks/migrate.php#L249) needs new code to figure out the next/previous version. But when jumping to a version that exists, this appears to work.

philsturgeon commented:
So this is a half fix and I should play with the rest? What is the advantage to what you've done? 

I'm not being offensive, I am just trying to work out what you've improved so I can continue, or see if I should just have a play or if I can provide some feedback.

gizzmo commented:
When i first started this, it was only to fix issues with migrating down, Mainly database version value and config value being out of sync when migrating down. Also i was under the impression (from looking at code) that versions were incremental, meaning that were never to have missing versions numbers.

philsturgeon commented:
Ok cool, I'll merge this then sort the rest. The non-incremental stuff is a new change so I'll take care of it separately. Thank you!
